File: tests/fixtures/slickrun3.scala
// Regression fixture for `agent/slickrun3`: the defects that stood between
// `tests/slick_run.sh`'s `p12_mapped` and slick's own query compiler.
//
// Every case here was found by running slick that scala-rs compiled, and each
// one is checked against real scalac 2.13.16 (see
// `crates/cli/tests/slickrun3.rs`, which also pins the emitted descriptors).

// --- 1. An abstract type member with a *compound* upper bound erases through
// nsc's `intersectionDominator`, and a class that narrows such a parameter
// owes the wide bridge. slick: `MappedColumnTypeFactory.base(…,
// BaseColumnType[U]): BaseColumnType[T]`, implemented by `MappedJdbcType` at
// `JdbcType`.
trait TypedType[T] { def name: String }
trait BaseTypedType[T] extends TypedType[T]

trait TypesComponent {
  type ColumnType[T] <: TypedType[T]
  type BaseColumnType[T] <: ColumnType[T] with BaseTypedType[T]

  trait ColumnTypeFactory {
    def base[T](tag: String, u: BaseColumnType[T]): BaseColumnType[T]
    def assertNonNull[A](t: BaseColumnType[A]): Unit =
      if (t == null) throw new NullPointerException("null column type")
  }

  val factory: ColumnTypeFactory

  // Reached through the interface, so the call uses the wide descriptor the
  // bridge has to answer.
  def viaFactory[T](tag: String, u: BaseColumnType[T]): String =
    factory.base(tag, u).name
}

trait JdbcType[T] extends BaseTypedType[T]

object JdbcComponent extends TypesComponent {
  type ColumnType[T] = JdbcType[T]
  type BaseColumnType[T] = JdbcType[T]

  object MappedJdbcType extends ColumnTypeFactory {
    // Narrows both the parameter and the result; nsc puts a
    // `(String, TypedType)TypedType` bridge on this class.
    def base[T](tag: String, u: JdbcType[T]): JdbcType[T] = {
      assertNonNull(u)
      new JdbcType[T] { def name = tag + "(" + u.name + ")" }
    }
  }
  val factory: ColumnTypeFactory = MappedJdbcType
  val intType: BaseColumnType[Int] = new JdbcType[Int] { def name = "int" }
}

// --- 2. `intersectionDominator` really does drop the other half of the bound.
// scala-reflect's `type TermName >: Null <: TermNameApi with Name` erases to
// the *interface* `TermNameApi`, which is not the abstract *class* `NameApi`
// that only `Name` brings in, so passing one where the other is expected needs
// a cast at the call site.
abstract class NameApi { def text: String }
trait TermNameApi

trait Names {
  type Name >: Null <: NameApi
  type TermName >: Null <: TermNameApi with Name
  def mkTerm(s: String): TermName
  def render(n: NameApi): String = "<" + n.text + ">"
}

object Universe extends Names {
  class NameImpl(val text: String) extends NameApi with TermNameApi
  type Name = NameImpl
  type TermName = NameImpl
  def mkTerm(s: String): TermName = new NameImpl(s)
}

// --- 3. A local's type is written in its own vocabulary: `map.infer(…)` is
// `map.Self`, not this class's `Self`. slick's `ResultSetMapping
// .withInferredType` destructured a tuple of one and `checkcast`ed a plain
// `Node` to a `ResultSetMapping`.
trait Node {
  type Self >: this.type <: Node
  def infer(depth: Int): Self
  def label: String
}
class Leaf extends Node {
  type Self = Leaf
  def infer(depth: Int): Self = this
  def label = "leaf"
}
class Mapping(val map: Node) extends Node {
  type Self = Mapping
  def infer(depth: Int): Self = this
  def label = "mapping"
  def inferred: String = {
    val (m2, d) = (map.infer(1), 7)
    m2.label + d
  }
}

// --- 4. A block in statement position discards its last expression, so a
// branching last expression is generated in statement mode too. slick's
// `QueryInterpolator.appendString` otherwise left one arm of an inner match
// with a value on the stack ("Inconsistent stackmap frames").
class Buf {
  var n = 0
  def +=(c: Char): Buf = { n += 1; this }
  def nl(): Unit = { n += 100 }
}

// --- 5. A `withFilter` whose result is *not* the receiver keeps its own type.
// slick's `ConstArray.withFilter(p): ConstArrayOp[T]`.
trait ArrayOp[T] { def foreach[R](f: T => R): Unit }
final class ConstArr[T](val xs: List[T]) {
  def foreach[R](f: T => R): Unit = xs.foreach(f)
  def withFilter(p: T => Boolean): ArrayOp[T] = new ArrayOp[T] {
    def foreach[R](f: T => R): Unit = xs.filter(p).foreach(f)
  }
}

// --- 6. `super.m` resolves to the first *concrete* `m`, never to a mixin that
// only re-declares it. slick's `BasicStreamingQueryActionExtensionMethodsImpl`
// narrows `result` covariantly and leaves it abstract, and the `$class` holder
// the call named does not exist at all.
trait Res { def show: String }
class ResA extends Res { def show = "A" }
trait Actions { def result: Res }
trait StreamingActions extends Actions { def result: ResA }
class Query0 extends Actions { def result: Res = new ResA }
class Query1 extends Query0 { override def result: ResA = new ResA }
class Query2 extends Query1 with StreamingActions {
  override def result: ResA = super.result
}

object Main {
  def append(b: Buf, str: String, skip: Boolean): Unit = {
    def appendString(s: String): Unit = {
      val len = s.length
      var pos = 0
      while (pos < len) {
        s.charAt(pos) match {
          case '\\' =>
            pos += 1
            if (pos < len) {
              s.charAt(pos) match {
                case c2 @ ('(' | ')') => if (!skip) b += c2
                case '{' =>
                  if (!skip) {
                    b += '('
                    b.nl()
                  }
                case 'n' => b.nl()
                case c2  => throw new IllegalArgumentException("bad " + c2)
              }
            }
          case c => b += c
        }
        pos += 1
      }
    }
    appendString(str)
  }

  def main(args: Array[String]): Unit = {
    println("base: " + JdbcComponent.viaFactory("mapped", JdbcComponent.intType))
    println("direct: " + JdbcComponent.MappedJdbcType.base[Int]("direct", JdbcComponent.intType).name)

    println("name: " + Universe.render(Universe.mkTerm("xs")))

    println("inferred: " + new Mapping(new Leaf).inferred)

    val b = new Buf
    append(b, "ab\\(cd\\{e\\nf", false)
    append(b, "\\(\\)", true)
    println("buf: " + b.n)

    val c = new ConstArr(List((1, "a"), (2, "b"), (3, "c")))
    for ((n, s: String) <- c) print(n.toString + s + " ")
    for (p <- c if p._1 > 1) print(p._2 + " ")
    println()

    println("super: " + new Query2().result.show)
  }
}

File: crates/cli/tests/slickrun3.rs
//! E2E tests for the `agent/slickrun3` slice (fixture prefix `slickrun3`).
//!
//! `tests/slick_run.sh`'s `p12_mapped` died on the very first line with
//!
//! ```text
//! NoSuchMethodError: slick.ast.TypedType
//!   RelationalTypesComponent$MappedColumnTypeFactory.base(
//!     Function1, Function1, ClassTag, slick.ast.TypedType)
//! ```
//!
//! Removing that took five defects, and the fifth left slick's whole query
//! compiler and SQL generator running on scala-rs-compiled class files,
//! producing SQL byte-identical to real scalac's build. Each one is a case in
//! `tests/fixtures/slickrun3.scala`:
//!
//! 1. **An abstract type member with a compound upper bound erased to
//!    `Object`.** SLS 3.7 erases it like a type parameter, and nsc reduces a
//!    compound bound with `Erasure.intersectionDominator` -- a parent that
//!    some other parent extends is shadowed, a real class beats a trait, and
//!    otherwise the first one wins. slick's `type BaseColumnType[T] <:
//!    ColumnType[T] & BaseTypedType[T]` is `slick.ast.TypedType`.
//!
//! 2. **A subclass that narrows such a parameter got no bridge.** After
//!    erasure nothing says `base(…, TypedType)` and `base(…, JdbcType)` are
//!    one method rather than two overloads, so `SymbolTable::
//!    erased_abstract_params` records which parameters were abstract *before*
//!    erasure and `gen::bridge_overrides` reads it. Without the bridge the
//!    interface method stayed abstract (`AbstractMethodError`).
//!
//! 3. **`intersectionDominator` drops information, and the call site pays for
//!    it.** scala-reflect's `type TermName >: Null <: TermNameApi with Name`
//!    erases to the *interface* `Names$TermNameApi` -- nsc's own `newTermName`
//!    returns exactly that -- which does not extend the abstract *class*
//!    `Names$NameApi`; only `Name` brings that in. Passing one where the other
//!    is expected needs a `checkcast` (`VerifyError: Type Names$TermNameApi is
//!    not assignable to Names$NameApi`, in the `ShapedValue` macro's
//!    expansion).
//!
//! 4. **A local's type was re-read through `this`.** `Typer::bind_found`
//!    applied `expand_type_members` to every identifier, so a local whose type
//!    mentioned `map.Self` was rebound to the enclosing class's `Self`. Only a
//!    *member* of a class is seen through a prefix. slick's
//!    `ResultSetMapping.withInferredType` destructured `(map2, newType)` and
//!    `checkcast`ed a plain `Node` to a `ResultSetMapping` -- every query the
//!    compiler ran.
//!
//! 5. **A block in statement position asked for its value.** `gen_stat` had no
//!    `Block` arm, so it fell through to `gen_expr` + pop and put a branching
//!    last expression back in value mode. slick's
//!    `QueryInterpolator.appendString` then generated an inner `match` for its
//!    `Any` lub and only the arms whose own type was not `Unit` left anything
//!    on the stack (`VerifyError: Inconsistent stackmap frames`).
//!
//! 6. **`withFilter`'s result was replaced by the receiver's type.** That rule
//!    exists for the collections, where the declared result is the receiver
//!    *widened*; it is wrong for a `withFilter` returning something else.
//!    slick's `ConstArray.withFilter(p): ConstArrayOp[T]` made the following
//!    `foreach` resolve to `ConstArray`'s and `checkcast` the anonymous
//!    `ConstArrayOp` to a `ConstArray`.
//!
//! 7. **`super.m` landed on a mixin that only re-declares `m`.** nsc resolves
//!    it to the first *concrete* member along the linearization. slick's
//!    `BasicStreamingQueryActionExtensionMethodsImpl` narrows `result`
//!    covariantly and leaves it abstract, and the `$class` holder the emitted
//!    `invokestatic` named does not exist at all
//!    (`NoClassDefFoundError`).
//!
//! Self-contained (own copies of the small helpers `e2e.rs` also has) per
//! `.agent-brief.md`: the shared test files belong to other in-flight agents.

use std::fs;
use std::path::PathBuf;
use std::process::Command;
use std::time::{SystemTime, UNIX_EPOCH};

fn bin() -> PathBuf {
    PathBuf::from(env!("CARGO_BIN_EXE_scala-rs"))
}

fn fixtures_dir() -> PathBuf {
    PathBuf::from(env!("CARGO_MANIFEST_DIR")).join("../../tests/fixtures")
}

fn tmp_dir(tag: &str) -> PathBuf {
    let nanos = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_nanos())
        .unwrap_or(0);
    static SEQ: std::sync::atomic::AtomicU32 = std::sync::atomic::AtomicU32::new(0);
    let seq = SEQ.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
    let p = std::env::temp_dir().join(format!(
        "scala-rs-slickrun3-{tag}-{}-{nanos}-{seq}",
        std::process::id()
    ));
    fs::create_dir_all(&p).unwrap();
    p
}

fn scala_library_jar() -> Option<PathBuf> {
    let cached = PathBuf::from("/tmp/scala-rs-lib/scala-library-2.13.16.jar");
    cached.is_file().then_some(cached)
}

fn java_available() -> bool {
    Command::new("java")
        .arg("-version")
        .output()
        .map(|o| o.status.success())
        .unwrap_or(false)
}

/// Compile the fixture against the real scala-library; returns the output dir.
fn compile_fixture() -> Option<PathBuf> {
    let jar = scala_library_jar()?;
    let out = tmp_dir("out");
    let res = Command::new(bin())
        .args([
            "compile",
            fixtures_dir().join("slickrun3.scala").to_str().unwrap(),
            "-d",
            out.to_str().unwrap(),
            "--scala-library",
            jar.to_str().unwrap(),
        ])
        .output()
        .expect("run scala-rs compile");
    assert!(
        res.status.success(),
        "scala-rs failed on slickrun3.scala:\n{}\n{}",
        String::from_utf8_lossy(&res.stdout),
        String::from_utf8_lossy(&res.stderr)
    );
    Some(out)
}

/// The whole fixture, run under the bytecode verifier, against real scalac
/// 2.13.16's own stdout for the same file.
#[test]
fn fixtures_slickrun3_matches_scalac() {
    if !java_available() {
        return;
    }
    let Some(jar) = scala_library_jar() else {
        eprintln!("skip slickrun3: no scala-library jar");
        return;
    };
    let out = compile_fixture().unwrap();
    let run = Command::new("java")
        .args([
            "-Xverify:all",
            "-cp",
            &format!("{}:{}", out.display(), jar.display()),
            "Main",
        ])
        .output()
        .expect("run the fixture");
    assert!(
        run.status.success(),
        "slickrun3 failed at run time:\n{}",
        String::from_utf8_lossy(&run.stderr)
    );
    let expected =
        fs::read_to_string(fixtures_dir().join("expected").join("slickrun3.txt")).unwrap();
    assert_eq!(String::from_utf8_lossy(&run.stdout), expected);
    let _ = fs::remove_dir_all(&out);
}

fn javap(out: &PathBuf, class: &str) -> Option<String> {
    let res = Command::new("javap")
        .args(["-p", "-cp", out.to_str().unwrap(), class])
        .output()
        .ok()?;
    res.status
        .success()
        .then(|| String::from_utf8_lossy(&res.stdout).into_owned())
}

/// The descriptors themselves, so a change that keeps the stdout by some other
/// route still has to say so out loud. Every expectation here is what real
/// scalac 2.13.16 emits for the same fixture.
#[test]
fn fixtures_slickrun3_erases_compound_bounds_like_nsc() {
    let Some(out) = compile_fixture() else {
        eprintln!("skip slickrun3 javap: no scala-library jar");
        return;
    };
    let Some(factory) = javap(&out, "TypesComponent$ColumnTypeFactory") else {
        eprintln!("skip slickrun3 javap: no javap");
        return;
    };
    // `BaseColumnType[T] <: ColumnType[T] with BaseTypedType[T]` erases to the
    // dominator `ColumnType`, hence to *its* bound `TypedType` -- not to
    // `Object`, and not to the shadowed-out `BaseTypedType`.
    assert!(
        factory.contains("TypedType base(java.lang.String, TypedType)"),
        "the interface should declare the wide `base`:\n{factory}"
    );
    assert!(
        factory.contains("void assertNonNull(TypedType)"),
        "and `assertNonNull` at the same erasure:\n{factory}"
    );

    let mapped = javap(&out, "JdbcComponent$MappedJdbcType$").unwrap();
    assert!(
        mapped.contains("JdbcType base(java.lang.String, JdbcType)"),
        "the narrow implementation:\n{mapped}"
    );
    assert!(
        mapped.contains("TypedType base(java.lang.String, TypedType)"),
        "and the bridge the interface call needs:\n{mapped}"
    );

    // The other half of a compound bound really is dropped: `TermName >: Null
    // <: TermNameApi with Name` is `TermNameApi`, exactly as nsc's own
    // `Names.newTermName` is declared in scala-reflect.jar.
    let universe = javap(&out, "Universe$").unwrap();
    assert!(
        universe.contains("TermNameApi mkTerm(java.lang.String)"),
        "`mkTerm`'s bridge should be at `TermNameApi`:\n{universe}"
    );
    let _ = fs::remove_dir_all(&out);
}
